### ISSUE TITLE
Pre-release 1.6: extra-deps produced by "solver" command no longer valid

### ISSUE THREAD

**Issue by paul-rouse** [commercialhaskell/stack#3533]
The `extra-deps` produced by `stack solver` can sometimes duplicate packages already specified in `packages` - this happens in the travis tests for [persistent](https://github.com/yesodweb/persistent) when attempting to use stack-1.6.  This did not cause any problem for stack-1.5.1, but `stack build` fails in pre-release 1.6.0.20171022 with messages like:

```
The same package name is used in multiple local packages

persistent used in:
- PLIndex "persistent-2.7.0"
- PLOther (PLFilePath "./persistent")
```

### Steps to reproduce

Unfortunately I think a fairly large setup is needed to show the problem:

* a multi-package setup
* dependencies between the packages
* released versions somewhere `stack solver` can find them

Example output is in [this travis log](https://travis-ci.org/paul-rouse/persistent/jobs/294804309).  At line 625 `stack solver` reports the extra-deps it is adding, and the error occurs at line 652.  The starting point in this case is this stack.yaml:

```
resolver: lts-8.2
packages:
  - ./persistent
  - ./persistent-template
  - ./persistent-sqlite
  - ./persistent-test
  - ./persistent-mongoDB
  - ./persistent-mysql
  - ./persistent-postgresql
  - ./persistent-redis
``` 

Most of the added `extra-deps` shown in the log are duplicates of these `packages`.

The new behaviour of detecting the duplication seems correct, but, if so, `solver` now needs to be more careful what it adds.

### Stack version

```
Version 1.6.0.20171022, Git revision 7bddfaf7f9f8cd9ec1c710fa83e77262e494eee4 (5285 commits) x86_64 hpack-0.18.1
```

**Comment by paul-rouse** [comment]
I found a somewhat simpler test case, but still involving a released package.  It seems that `solver` does not recognise when the local package *A* satisfies a dependency in *B* which requires a later version than the one in the snapshot. 

```
#!/bin/bash

mkdir workingdir
cd workingdir

# The stack version to test
curl -L 'https://github.com/commercialhaskell/stack/releases/download/v1.6.0.20171022/stack-1.6.0.20171022-linux-x86_64-static.tar.gz' | tar xz --wildcards --strip-components=1 '*/stack'

# I have a ghc-7.10.3 lying around, so I'll use that instead of --install-ghc
PATH=/opt/ghc/7.10.3/bin:$PATH
STACK="./stack --system-ghc --no-install-ghc"

$STACK --version

# AFAIK the only important point about the choice of package is that its
# latest version (both released and in the repo) is newer than the one in
# the snapshot we are going to use.
git clone https://github.com/paul-rouse/yesod-auth-hashdb.git

$STACK new test simple

# Add a dependency to a NEWER version than the one in the snapshot we will use
echo "                    , yesod-auth-hashdb >= 1.6" >> test/test.cabal

# Just to be certain which stack.yaml we use
rm yesod-auth-hashdb/stack.yaml test/stack.yaml

cat > stack.yaml <<EOF
resolver: lts-6.35
packages:
  - ./yesod-auth-hashdb
  - ./test
EOF

$STACK --resolver lts-6.35 solver --update-config

# Bang!
$STACK build
```

**Comment by snoyberg** [comment]
Thanks for the report, I've added a new failing test case in 9378c3768cc08142d6a55ba05ca2a0f1271ba5b8. Now to see if I can fix it :)

**Comment by snoyberg** [comment]
OK, this should be resolved by #3536. If you could test that out, it would be much appreciated.

**Comment by paul-rouse** [comment]
Yes, that is working perfectly, both in the test script shown above, and in my copy of the `persistent` tests.  Thank you!  I think this can be closed now, since it is merged.
